edisgo/flex_opt/reinforce_grid.py

```diff
@@ -967,3 +979,168 @@ def enhanced_reinforce_grid(
             edisgo_obj.results.measures = msg
 
     return edisgo_obj
+
+
+def run_separate_lv_grids(edisgo_obj: EDisGo, threshold: int | float = 2) -> None:
+    """
+    Separate all highly overloaded LV grids within the MV grid.
+
+    The loading is approximated by aggregation of all load and generator time series
+    and comparison with the total nominal apparent power of the MV/LV transformer(s).
+    This approach is chosen because this method is aims at resolving highly overloaded
+    grid situations in which cases the power flow often does not converge. This method
+    ignores grid losses and voltage deviations. Originating and new LV grids can be
+    separated multiple times if the overloading is very high.
+
+    Parameters
+    ----------
+    edisgo_obj : :class:`~.EDisGo`
+    threshold : int or float
+        Overloading threshold. If the overloading is higher than the threshold times
+        the total nominal apparent power of the MV/LV transformer(s) the grid is
+        separated.
+
+    Returns
+    -------
+    :class:`~.EDisGo`
+        The reinforced eDisGo object.
+    """
+    lv_grids = list(edisgo_obj.topology.mv_grid.lv_grids)
+    n_grids_init = len(lv_grids)
+
+    first_run = True
+
+    active_str = "{}_active_power"
+    reactive_str = "{}_reactive_power"
+    tech_str = "{}_df"
+    techs = ["generators", "loads", "storage_units"]
+
+    n = 0
+    max_iterations = 100
+
+    while (
+        n_grids_init != len(list(edisgo_obj.topology.mv_grid.lv_grids)) or first_run
+    ) and n < max_iterations:
+        n += 1
+        first_run = False
+
+        lv_grids = list(edisgo_obj.topology.mv_grid.lv_grids)
+        n_grids_init = len(lv_grids)
+
+        for lv_grid in lv_grids:
+            tech_dict = {}
+
+            for tech in techs:
+                units = getattr(lv_grid, tech_str.format(tech)).index
+                active_power = (
+                    getattr(
+                        edisgo_obj.timeseries,
+                        active_str.format(tech),
+                    )
+                    .loc[:, units]
+                    .astype(float)
+                )
+
+                reactive_power = (
+                    getattr(
+                        edisgo_obj.timeseries,
+                        reactive_str.format(tech),
+                    )
+                    .loc[:, units]
+                    .astype(float)
+                )
+
+                if tech == "storage_units":
```

Comment:
I am not quite sure if storage units are handled correctly like this.. Wouldn't it be easier to take np.hypot after adding up all active and reactive powers of loads, generators and storage units?

Reply:
Yeah, you are right. Somehow my thinking was far too complicated. I changed it now:

https://github.com/openego/eDisGo/blob/f473b5ac6990fe55a5b46f1939769ff296d2b6c3/edisgo/flex_opt/reinforce_grid.py#L1031-L1069